Panic in hclwrite parser: didn't find any token of type TokenIdent

Originally reported over in apparentlymart/terraform-clean-syntax#3 by @mattsawyer77.

On certain input, it seems that the analysis of traversals in expressions fails and produces a panic:

```
panic: didn't find any token of type TokenIdent

goroutine 1 [running]:
github.com/hashicorp/hcl/v2/hclwrite.inputTokens.PartitionType(0xc00036be20, 0x0, 0x0, 0xc0003c5fd8, 0x0, 0x0, 0x49, 0x0, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:88 +0x580
github.com/hashicorp/hcl/v2/hclwrite.inputTokens.PartitionTypeSingle(0xc00036be20, 0x0, 0x0, 0xc0003c5fd8, 0x0, 0x0, 0xc000000049, 0x0, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:92 +0xe3
github.com/hashicorp/hcl/v2/hclwrite.parseTraversalStep(0x130e460, 0xc000298050, 0xc00036be20, 0x0, 0x0, 0xc0003c5fd8, 0x0, 0x0, 0xc00036be20, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:395 +0x364
github.com/hashicorp/hcl/v2/hclwrite.parseTraversal(0xc000263280, 0x3, 0x4, 0xc00036be20, 0x0, 0x0, 0xc0003c5fd8, 0x0, 0x0, 0xc00036be20, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:377 +0x3da
github.com/hashicorp/hcl/v2/hclwrite.parseExpression(0x130faa0, 0xc0000b6700, 0xc00036be20, 0x2, 0x2, 0xc0003c5fd8, 0x2, 0x2, 0x18b)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:356 +0x175
github.com/hashicorp/hcl/v2/hclwrite.parseAttribute(0xc00028bea0, 0xc00036ba60, 0x0, 0x0, 0xc0003c5f88, 0x0, 0x0, 0xc00036b9a0, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:260 +0xaf4
github.com/hashicorp/hcl/v2/hclwrite.parseBodyItem(0x130cc20, 0xc00028bea0, 0xc00036b9a0, 0x17, 0x17, 0xc0003c5f78, 0x17, 0x17, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:217 +0x632
github.com/hashicorp/hcl/v2/hclwrite.parseBody(0xc0000b5a20, 0xc00036af20, 0x33, 0x33, 0xc0003c5e98, 0x33, 0x33, 0x0, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:193 +0x68d
github.com/hashicorp/hcl/v2/hclwrite.parseBlock(0xc000286b40, 0xc00036af20, 0x0, 0x0, 0xc0003c5e98, 0x0, 0x0, 0xc00036aa40, 0x5, 0x5, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:329 +0xf72
github.com/hashicorp/hcl/v2/hclwrite.parseBodyItem(0x130cc60, 0xc000286b40, 0xc00036a9e0, 0x70, 0x70, 0xc0003c5e28, 0x70, 0x70, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:219 +0x4f7
github.com/hashicorp/hcl/v2/hclwrite.parseBody(0xc0000b5b80, 0xc00033c000, 0x836, 0xa00, 0xc0003c2000, 0x836, 0x836, 0x0, 0x0, 0x0, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:193 +0x68d
github.com/hashicorp/hcl/v2/hclwrite.parse(0xc000102000, 0x38eb, 0x3aeb, 0xc0000e41e0, 0x1c, 0x1, 0x1, 0x0, 0xc0000c3c48, 0xc000000000, ...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/parser.go:53 +0x2e4
github.com/hashicorp/hcl/v2/hclwrite.ParseConfig(...)
        /Users/sawyer/gocode/pkg/mod/github.com/hashicorp/hcl/v2@v2.1.0/hclwrite/public.go:27
main.processFile(0xc0000e41e0, 0x1c, 0x1a4)
        /Users/sawyer/terraform-clean-syntax/main.go:75 +0xc8
main.processItem(0xc0000e41e0, 0x1c)
        /Users/sawyer/terraform-clean-syntax/main.go:52 +0x25d
main.processDir(0x7ffeefbff63f, 0x16)
        /Users/sawyer/terraform-clean-syntax/main.go:64 +0x119
main.processItem(0x7ffeefbff63f, 0x16)
        /Users/sawyer/terraform-clean-syntax/main.go:44 +0x17b
main.main()
        /Users/sawyer/terraform-clean-syntax/main.go:30 +0xca
```

It looks like the relevant bit of code is here-ish:

https://github.com/hashicorp/hcl/blob/7f9aa845c107b03e33729886a8b5c5f4be74766a/hclwrite/parser.go#L391-L400

This codepath is assuming that any traversal step marked by the main parser as being a root (the beginning of a traversal) or an attribute (something that comes after a `.`) will contain an identifier in it. On the surface that seems like a reasonable assumption, but apparently there is at least one exceptional case where that isn't true.

--- a/hclsyntax/expression.go
+++ b/hclsyntax/expression.go
@@ -615,8 +615,9 @@ type IndexExpr struct {
 	Collection Expression
 	Key        Expression
 
-	SrcRange  hcl.Range
-	OpenRange hcl.Range
+	SrcRange     hcl.Range
+	OpenRange    hcl.Range
+	BracketRange hcl.Range
 }
 
 func (e *IndexExpr) walkChildNodes(w internalWalkFunc) {
@@ -631,7 +632,7 @@ func (e *IndexExpr) Value(ctx *hcl.EvalContext) (cty.Value, hcl.Diagnostics) {
 	diags = append(diags, collDiags...)
 	diags = append(diags, keyDiags...)
 
-	val, indexDiags := hcl.Index(coll, key, &e.SrcRange)
+	val, indexDiags := hcl.Index(coll, key, &e.BracketRange)
 	setDiagEvalContext(indexDiags, e, ctx)
 	diags = append(diags, indexDiags...)
 	return val, diags

--- a/hclsyntax/parser.go
+++ b/hclsyntax/parser.go
@@ -760,7 +760,7 @@ Traversal:
 					Each:   travExpr,
 					Item:   itemExpr,
 
-					SrcRange:    hcl.RangeBetween(dot.Range, lastRange),
+					SrcRange:    hcl.RangeBetween(from.Range(), lastRange),
 					MarkerRange: hcl.RangeBetween(dot.Range, marker.Range),
 				}
 
@@ -819,7 +819,7 @@ Traversal:
 					Each:   travExpr,
 					Item:   itemExpr,
 
-					SrcRange:    hcl.RangeBetween(open.Range, travExpr.Range()),
+					SrcRange:    hcl.RangeBetween(from.Range(), travExpr.Range()),
 					MarkerRange: hcl.RangeBetween(open.Range, close.Range),
 				}
 
@@ -867,8 +867,9 @@ Traversal:
 						Collection: ret,
 						Key:        keyExpr,
 
-						SrcRange:  rng,
-						OpenRange: open.Range,
+						SrcRange:     hcl.RangeBetween(from.Range(), rng),
+						OpenRange:    open.Range,
+						BracketRange: rng,
 					}
 				}
 			}
@@ -899,7 +900,7 @@ func makeRelativeTraversal(expr Expression, next hcl.Traverser, rng hcl.Range) E
 		return &RelativeTraversalExpr{
 			Source:    expr,
 			Traversal: hcl.Traversal{next},
-			SrcRange:  rng,
+			SrcRange:  hcl.RangeBetween(expr.Range(), rng),
 		}
 	}
 }

--- a/hclsyntax/parser_test.go
+++ b/hclsyntax/parser_test.go
@@ -1834,12 +1834,16 @@ block "valid" {}
 									},
 								},
 								SrcRange: hcl.Range{
-									Start: hcl.Pos{Line: 1, Column: 26, Byte: 25},
+									Start: hcl.Pos{Line: 1, Column: 8, Byte: 7},
 									End:   hcl.Pos{Line: 1, Column: 39, Byte: 38},
 								},
 								OpenRange: hcl.Range{
 									Start: hcl.Pos{Line: 1, Column: 26, Byte: 25},
 									End:   hcl.Pos{Line: 1, Column: 27, Byte: 26},
+								},
+								BracketRange: hcl.Range{
+									Start: hcl.Pos{Line: 1, Column: 26, Byte: 25},
+									End:   hcl.Pos{Line: 1, Column: 39, Byte: 38},
 								},
 							},
 							SrcRange: hcl.Range{
@@ -1869,6 +1873,91 @@ block "valid" {}
 				EndRange: hcl.Range{
 					Start: hcl.Pos{Line: 2, Column: 1, Byte: 41},
 					End:   hcl.Pos{Line: 2, Column: 1, Byte: 41},
+				},
+			},
+		},
+		{
+			"a = \"${var.public_subnets[*]}\"\n",
+			0,
+			&Body{
+				Attributes: Attributes{
+					"a": {
+						Name: "a",
+						Expr: &TemplateWrapExpr{
+							Wrapped: &SplatExpr{
+								Source: &ScopeTraversalExpr{
+									Traversal: hcl.Traversal{
+										hcl.TraverseRoot{
+											Name: "var",
+
+											SrcRange: hcl.Range{
+												Start: hcl.Pos{Line: 1, Column: 8, Byte: 7},
+												End:   hcl.Pos{Line: 1, Column: 11, Byte: 10},
+											},
+										},
+										hcl.TraverseAttr{
+											Name: "public_subnets",
+
+											SrcRange: hcl.Range{
+												Start: hcl.Pos{Line: 1, Column: 11, Byte: 10},
+												End:   hcl.Pos{Line: 1, Column: 26, Byte: 25},
+											},
+										},
+									},
+
+									SrcRange: hcl.Range{
+										Start: hcl.Pos{Line: 1, Column: 8, Byte: 7},
+										End:   hcl.Pos{Line: 1, Column: 26, Byte: 25},
+									},
+								},
+								Each: &AnonSymbolExpr{
+									SrcRange: hcl.Range{
+										Start: hcl.Pos{Line: 1, Column: 26, Byte: 25},
+										End:   hcl.Pos{Line: 1, Column: 29, Byte: 28},
+									},
+								},
+								Item: &AnonSymbolExpr{
+									SrcRange: hcl.Range{
+										Start: hcl.Pos{Line: 1, Column: 26, Byte: 25},
+										End:   hcl.Pos{Line: 1, Column: 29, Byte: 28},
+									},
+								},
+								SrcRange: hcl.Range{
+									Start: hcl.Pos{Line: 1, Column: 8, Byte: 7},
+									End:   hcl.Pos{Line: 1, Column: 29, Byte: 28},
+								},
+								MarkerRange: hcl.Range{
+									Start: hcl.Pos{Line: 1, Column: 26, Byte: 25},
+									End:   hcl.Pos{Line: 1, Column: 29, Byte: 28},
+								},
+							},
+							SrcRange: hcl.Range{
+								Start: hcl.Pos{Line: 1, Column: 5, Byte: 4},
+								End:   hcl.Pos{Line: 1, Column: 31, Byte: 30},
+							},
+						},
+						SrcRange: hcl.Range{
+							Start: hcl.Pos{Line: 1, Column: 1, Byte: 0},
+							End:   hcl.Pos{Line: 1, Column: 31, Byte: 30},
+						},
+						NameRange: hcl.Range{
+							Start: hcl.Pos{Line: 1, Column: 1, Byte: 0},
+							End:   hcl.Pos{Line: 1, Column: 2, Byte: 1},
+						},
+						EqualsRange: hcl.Range{
+							Start: hcl.Pos{Line: 1, Column: 3, Byte: 2},
+							End:   hcl.Pos{Line: 1, Column: 4, Byte: 3},
+						},
+					},
+				},
+				Blocks: Blocks{},
+				SrcRange: hcl.Range{
+					Start: hcl.Pos{Line: 1, Column: 1, Byte: 0},
+					End:   hcl.Pos{Line: 2, Column: 1, Byte: 31},
+				},
+				EndRange: hcl.Range{
+					Start: hcl.Pos{Line: 2, Column: 1, Byte: 31},
+					End:   hcl.Pos{Line: 2, Column: 1, Byte: 31},
 				},
 			},
 		},
@@ -2353,13 +2442,18 @@ block "valid" {}
 							},
 							SrcRange: hcl.Range{
 								Filename: "",
-								Start:    hcl.Pos{Line: 1, Column: 30, Byte: 29},
+								Start:    hcl.Pos{Line: 1, Column: 5, Byte: 4},
 								End:      hcl.Pos{Line: 1, Column: 43, Byte: 42},
 							},
 							OpenRange: hcl.Range{
 								Filename: "",
 								Start:    hcl.Pos{Line: 1, Column: 30, Byte: 29},
 								End:      hcl.Pos{Line: 1, Column: 31, Byte: 30},
+							},
+							BracketRange: hcl.Range{
+								Filename: "",
+								Start:    hcl.Pos{Line: 1, Column: 30, Byte: 29},
+								End:      hcl.Pos{Line: 1, Column: 43, Byte: 42},
 							},
 						},
 						SrcRange: hcl.Range{

--- a/hclwrite/parser_test.go
+++ b/hclwrite/parser_test.go
@@ -555,6 +555,150 @@ func TestParse(t *testing.T) {
 				},
 			},
 		},
+		{
+			"a = foo[bar]\n",
+			TestTreeNode{
+				Type: "Body",
+				Children: []TestTreeNode{
+					{
+						Type: "Attribute",
+						Children: []TestTreeNode{
+							{
+								Type: "comments",
+							},
+							{
+								Type: "identifier",
+								Val:  "a",
+							},
+							{
+								Type: "Tokens",
+								Val:  " =",
+							},
+							{
+								Type: "Expression",
+								Children: []TestTreeNode{
+									{
+										Type: "Traversal",
+										Children: []TestTreeNode{
+											{
+												Type: "TraverseName",
+												Children: []TestTreeNode{
+													{
+														Type: "identifier",
+														Val:  " foo",
+													},
+												},
+											},
+										},
+									},
+									{
+										Type: "Tokens",
+										Val:  "[",
+									},
+									{
+										Type: "Traversal",
+										Children: []TestTreeNode{
+											{
+												Type: "TraverseName",
+												Children: []TestTreeNode{
+													{
+														Type: "identifier",
+														Val:  "bar",
+													},
+												},
+											},
+										},
+									},
+									{
+										Type: "Tokens",
+										Val:  "]",
+									},
+								},
+							},
+							{
+								Type: "comments",
+							},
+							{
+								Type: "Tokens",
+								Val:  "\n",
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			"a = foo[bar].baz\n",
+			TestTreeNode{
+				Type: "Body",
+				Children: []TestTreeNode{
+					{
+						Type: "Attribute",
+						Children: []TestTreeNode{
+							{
+								Type: "comments",
+							},
+							{
+								Type: "identifier",
+								Val:  "a",
+							},
+							{
+								Type: "Tokens",
+								Val:  " =",
+							},
+							{
+								Type: "Expression",
+								Children: []TestTreeNode{
+									{
+										Type: "Traversal",
+										Children: []TestTreeNode{
+											{
+												Type: "TraverseName",
+												Children: []TestTreeNode{
+													{
+														Type: "identifier",
+														Val:  " foo",
+													},
+												},
+											},
+										},
+									},
+									{
+										Type: "Tokens",
+										Val:  "[",
+									},
+									{
+										Type: "Traversal",
+										Children: []TestTreeNode{
+											{
+												Type: "TraverseName",
+												Children: []TestTreeNode{
+													{
+														Type: "identifier",
+														Val:  "bar",
+													},
+												},
+											},
+										},
+									},
+									{
+										Type: "Tokens",
+										Val:  "].baz",
+									},
+								},
+							},
+							{
+								Type: "comments",
+							},
+							{
+								Type: "Tokens",
+								Val:  "\n",
+							},
+						},
+					},
+				},
+			},
+		},
 	}
 
 	for _, test := range tests {

